function/pkg/workspace/workspace.go

```diff
@@ -58,6 +83,83 @@ func fromRuntime(runtime types.Runtime) (workspace, error) {
 	}
 }
 
+func Synchronise(ctx context.Context, config Cfg, outputPath string, build client.Build) error {
+	return synchronise(ctx, config, outputPath, build, defaultWriterProvider)
+}
+
+func synchronise(ctx context.Context, config Cfg, outputPath string, build client.Build, writerProvider WriterProvider) error {
```

Comment:
I think that this function could use some refactoring. The whole synchronization logic is in this single function and I'm not even sure if it belongs in the `workspace` package.